src/swupdd-main.c

```diff
@@ -22,751 +18,765 @@
-                                   GDBusMethodInvocation *invocation,
-                                   GVariant *arg_options,
-                                   const gchar *arg_bundle,
-                                   gpointer user_data)
+static int on_childs_output(sd_event_source *s, int fd, uint32_t revents, void *userdata)
 {
-    daemon_state_t *dstate = (daemon_state_t *) user_data;
-    GError *error = NULL;
-    GList *args = NULL;
-    GVariantIter iter;
-    GVariant *value;
-    gchar *key;
-    gchar **argv;
-
-    g_message ("bundleRemove");
-    clr_software_update__complete_bundle_remove (object, invocation,
-                                                 (dstate->process == NULL));
-    if (dstate->process)
-        return FALSE;
-
-    args = g_list_append (args, g_strdup (SWUPD_CLIENT));
-    args = g_list_append (args, g_strdup ("bundle-remove"));
-
-    g_variant_iter_init (&iter, arg_options);
-    while (g_variant_iter_next (&iter, "{sv}", &key, &value))
-    {
-        _handle_common_options (key, value, args);
-        g_free (key);
-        g_variant_unref (value);
-    }
-
-    if (arg_bundle)
-        args = g_list_append (args,
-                              g_strdup (arg_bundle));
-
-    args = _set_defaults (args, dstate);
-    argv = _list_to_strv (args);
-    dstate->method = "bundleRemove";
-    dstate->process = g_subprocess_newv ((const gchar * const *) argv,
-                                         G_SUBPROCESS_FLAGS_STDOUT_PIPE |
-                                         G_SUBPROCESS_FLAGS_STDERR_MERGE,
-                                         &error);
-    if (dstate->process == NULL)
-    {
-        gchar *errmsg[2] = { NULL, NULL };
-        errmsg[0] = error->message;
-        clr_software_update__emit_request_completed (object,
-                                                     dstate->method,
-                                                     -1,
-                                                     (const gchar * const *) errmsg);
-        g_error_free (error);
-    }
-    else
-    {
-        g_subprocess_communicate_utf8_async (dstate->process,
-                                             NULL,
-                                             NULL,
-                                             _on_process_done,
-                                             dstate);
-    }
-
-    g_free (argv);
-    g_list_free_full (args, g_free);
-
-    return TRUE;
+	daemon_state_t *context = userdata;
+	int r = 0;
+	char buffer[PIPE_BUF + 1];
+	ssize_t count;
+	count = read(fd, buffer, PIPE_BUF);
+	if (count > 0) {
+		fwrite(buffer, 1, count, stdout);
+		fflush(stdout);
+		buffer[count] = '\0';
+		r = sd_bus_emit_signal(context->bus,
+				       "/org/O1/swupdd/Client",
+				       "org.O1.swupdd.Client",
+				       "childOutputReceived", "s", buffer);
+		if (r < 0) {
+			ERR("Failed to emit signal: %s", strerror(-r));
+		}
+	} else if (count < 0) {
```

Comment:
> it may be `EINTR` or `EAGAIN` (since you should set the fd to non-blocking), these you should skip the error and try again later.

I didn't use `O_NONBLOCK` because I don't fully understand its semantic in this case. If the child program (which we can't control by means other than signals) fails to write to `stdout` it won't try to write again: there's no code in swupd-client which checks return values of `printf()`. Which means we'll just miss some potentially important output.

On the reading side I always expect the fd ready to be read because that's the reason we receive an IO event in first place. I assume that receiving an IO event without actual data in the buffer would be an error.